narwhal/worker/src/batch_fetcher.rs

```diff
@@ -0,0 +1,530 @@
+                    .worker_batch_fetch
+                    .with_label_values(&["local", "success"])
+                    .inc();
+                fetched_batches.insert(digest, batch);
+            } else {
+                self.metrics
+                    .worker_batch_fetch
+                    .with_label_values(&["local", "failure"])
+                    .inc();
+            }
+        }
+
+        fetched_batches
+    }
+
+    /// This future performs a fetch from a given remote worker
+    /// This future performs infinite retries with exponential backoff
+    /// You can specify stagger_delay before request is issued
+    async fn fetch_remote(
+        &self,
+        stagger_delay: Duration,
+        worker: NetworkPublicKey,
+        digests: HashSet<BatchDigest>,
+    ) -> HashMap<BatchDigest, Batch> {
+        tokio::time::sleep(stagger_delay).await;
+        // TODO: Make these config parameters
+        let timeout = Duration::from_secs(10);
+        let mut attempt = 0usize;
+        let mut fetched_batches: HashMap<BatchDigest, Batch> = HashMap::new();
+        loop {
+            attempt += 1;
+            debug!(
+                "Remote attempt #{attempt} to fetch {} digests from {worker}",
+                digests.len(),
+            );
+            let request_batch_guard =
+                PendingGuard::make_inc(&self.metrics.pending_remote_request_batch);
+            let response = self
+                .safe_request_batches(digests.clone(), worker.clone(), timeout)
+                .await;
+            drop(request_batch_guard);
+            match response {
+                Ok(remote_batches) => {
+                    self.metrics
+                        .worker_batch_fetch
+                        .with_label_values(&["remote", "success"])
+                        .inc();
+                    debug!("Found {} batches remotely", remote_batches.len());
+                    for (remote_batch_digest, remote_batch) in remote_batches {
+                        if digests.contains(&remote_batch_digest) {
+                            fetched_batches.insert(remote_batch_digest, remote_batch);
+                        }
+                    }
+                    return fetched_batches;
+                }
+                Err(err) => {
+                    if err.to_string().contains("Timeout") {
+                        self.metrics
+                            .worker_batch_fetch
+                            .with_label_values(&["remote", "timeout"])
+                            .inc();
+                    } else {
+                        self.metrics
+                            .worker_batch_fetch
+                            .with_label_values(&["remote", "fail"])
+                            .inc();
+                    }
+                    debug!("Error retrieving payloads {digests:?} from {worker} attempt {attempt}: {err}")
+                }
+            }
+            tokio::time::sleep(timeout).await;
```

Comment:
This would mean we would sleep 10s after the attempt no matter what. How is this better than the exponential backoff code from before?

```
let deadline = Instant::now() + timeout;
...
timeout += timeout / 2;
timeout = std::cmp::min(max_timeout, timeout);
// Since the call might have returned before timeout, we wait until originally planned deadline
tokio::time::sleep_until(deadline).await;
```

Reply:
Keeping the existing backoff now.